Kafka: Add support for the delete-records API

graphcareful opened the pull request:
This PR adds the delete-records feature to redpanda. This feature allows a kafka client to truncate data from a topic/partition at a given offset. To support this in redpanda two non-trivial tasks were implemented:

1. Addition of new stm (and corresponding special batch type) to process prefix truncation events.
2. Addition of new concept (visible_start_offset) which is a delta into the first segment of the log representing the first offset a fetch request could read from.

Note, as of this branch delete-records requests for cloud enabled topics will be rejected. Work on supporting this is slated to be completed in a follow up PR, ~~which depends on a currently still in review PR~~ #9994  

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

### Features

* Adds support for the delete-records API

graphcareful commented:
Added more ducktape tests

emaxerrno commented:
wondering if franz-go has some delete records test, or if the librdkafka test suite should be enabled for this or some other external-API project that exercises this path for edge case detection.

jcsp commented on tests/rptest/tests/prefix_truncate_test.py:
```diff
@@ -0,0 +1,201 @@
+from rptest.services.storage import Segment
+
+
+class PrefixTruncateTest(RedpandaTest):
```
Let's call this DeleteRecordsTest (+ similarly name test cases) -- prefix truncation is a more general concept that we do in other situations as well.

jcsp commented on tests/rptest/tests/prefix_truncate_test.py:
```diff
@@ -0,0 +1,201 @@
+                return int(seg.base_index.split('-')[0])
+            if seg.data_file is not None:
+                return int(seg.data_file.split('-')[0])
+            return 0
```
What is the valid condition where a segment from the python scan of local storage would have no base index?  This feels like more of an error case.

jcsp commented on tests/rptest/tests/prefix_truncate_test.py:
```diff
@@ -0,0 +1,201 @@
+                return int(seg.data_file.split('-')[0])
+            return 0
+
+        max_base_index = max([to_base_index(seg) for seg in snapshot[leader]])
```
This helper seems to assume that a DeleteRecords call is always deleting all but the last segment (it's calculating 
the lower bound as the maximum of segment lower bounds).

It is necessary to also test other cases, like doing a DeleteRecords in the middle of the data range, doing one before the start of the data range.  Otherwise DeleteRecords could just always be nuking everything but the last segment, and this test would pass.

graphcareful replied:
Tests for this case coming in next force push , plus other edge case scenarios as well

jcsp commented on tests/rptest/util.py:
```diff
@@ -200,15 +200,16 @@ def segments_removed():
         return True
 
     wait_until(segments_removed,
-               timeout_sec=180,
```
Why is it correct to decrease this timeout?  I suspect some of the failures in other tests may be the result of this change.

graphcareful replied:
Nice catch, this wasn't meant to be committed

jcsp commented on src/v/cluster/partition.h:
```diff
@@ -86,6 +86,16 @@ class partition {
     }
 
+    /**
+     * The first index at which a consumer could read from. This will differ
```
Can you give an example of when the start_offset from a retention-driven prefix truncation would differ from the visible_start_offset from a DeleteRecords?  It's not obvious to me why they need to be separate concepts, couldn't we just have one start offset that means "data below this is not readable", and use it for both cases?

graphcareful replied:
@mmaslankaprv would there be a valid need to be able to read from before the newly proposed `visible_start_offset()` ? If theres not, then Johns suggestion makes sense and I should just modify how `start_offset()` works

graphcareful replied:
This has been implemented: https://github.com/redpanda-data/redpanda/compare/187597f7ca3963a1003a4c70c36cfb942c5d535b..1189f7cc50cde38a366c9317a6b9ed3ea298c8d8

jcsp commented on src/v/features/feature_table.h:
```diff
@@ -57,6 +57,7 @@ enum class feature : std::uint64_t {
     membership_change_controller_cmds = 1ULL << 22U,
     controller_snapshots = 1ULL << 23U,
     cloud_storage_manifest_format_v2 = 1ULL << 24U,
+    log_prefix_truncation = 1ULL << 25U,
```
This naming is confusing: prefix truncation is not new, this particular mechanism is.  If you want to avoid calling it delete_records, then something like "explicit_prefix_truncation" might make more sense.  But it would also be fine just to call this "delete_records".

jcsp commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -145,6 +145,11 @@ static ss::future<result_t> prefix_truncate(
+        } else if (errc == cluster::errc::feature_disabled) {
+            vlog(
+              klog.warn,
+              "Prefix truncate request rejected - feature not yet activated");
```
I don't think this message would make sense to a user: should say "DeleteRecords" so that they know what kind of external request this is talking about.

jcsp commented on tests/rptest/tests/prefix_truncate_test.py:
```diff
@@ -0,0 +1,201 @@
+
+class PrefixTruncateTest(RedpandaTest):
+    """
+    The purpose of this test is to exercise the delete-records API which
```
Other cases to think about:
- DeleteRecords of data which is below max_collectible_offset
- DeleteRecords ahead of hwm
- DeleteRecords below lwm
- DeleteRecords on a partition that doesn't exist
- DeleteRecords on a read replica topic
- DeleteRecords on an empty topic

jcsp commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -16,23 +16,140 @@
+    if (now >= deadline) {
+        co_return errc::timeout;
+    }
+    const auto timeout = ss::lowres_clock::now() + (deadline - now);
```
This looks redundant.  If `now()` is equal to `now) then the "timeout" is just equal to `deadline`.

jcsp reviewed (commented):
Looks like lots of tests are  failing, perhaps related to the timeout change in the PR.

graphcareful commented:
Changes in [force-push]
- Modify `prefix_truncate` name schema to `delete_records`
- Fix bug where truncating with -1 (high watermark) may cause a crash
- Fix bug that made offset translation not work correctly due to not excluding the new special batch type in translation accounting
- Fix bug that made request return early due to off-by-one error in firing new min offset event notification
- Fix bug that would made subsequent request to prefix truncate within the same segment fail due to not updating conditionals in `consensus::write_snapshot`
- Fix bug that would cause a crash when writing a snapshot at the first segment. `consensus::write_snapshot` expects `get_term` to always work, which was previously true since `write_snapshot` would never be called at a `last_included_index` value of `0`. Now this is not the case since `write_snapshot` can be called just to increment the new log start delta.
- Fix bug where `log_eviction_stm` would start processing events from `start_offset()` instead of `visible_start_offset()`. No issues arose because events would be ignored, after correctly determining that the older events aren't applicable to a log with a `start_offset + delta` ahead of the requested.
- Fix bug where delete-records at the `high watermark - 1` offset (truncate all data) failed due to off-by-one error
- Fix bug where `log_eviction_stm` was not truncating at segment end offsets but rather segment start offsets
- Modify KCL::delete_records for better error condition handling
- Much more ducktape tests

graphcareful commented:
Leaving a comment here about the proposed testing plan. All testing will be done in Ducktape, desired scenarios to test are:
- [  ] Test moving delta of current segment
- [  ] Test moving delta of current segment multiple times in a row
- [  ] Test deletion of data at segment boundary
- [  ] Test deletion of data within a segment, between boundaries
- [  ] Test truncation at offsets 0, < 0 and >= high watermark all result in out of bounds errors
- [  ] Test correct errors returned on topics/partitions that don't exist
- [  ] Test truncation at high watermark (-1) deletes all data
- [  ] Test an edge case situations where begin, end are very close like [38, 40)
- [  ] Test all of the above parameterized with cloud storage enabled topics
- [  ] Test scenarios where node failures are introduced (list scenarios - in-progress)

graphcareful commented:
Changes in force-push[https://github.com/redpanda-data/redpanda/compare/1626948815c0e9e0358e676094613f841a125e69..ebb4b29487e9c1de0dc3a5d5b811b34845800879]
- Rebase against dev to resolve conflicts in feature_table (another change added a new feature)
- Modified log_eviction housekeeping loop to periodically sleep when there is no work to do, or in the case the work item was re-enqueued due to the offset being greater then the max collectible.
- Fixed a bug that was resulting in test failures incorrect variable was being referred to in `log_eviction_stm::do_process_log_eviction` 
- Fixed a newly written flaky ducktape test by repeating if `NOT_LEADER` is observed 
- Made some changes around previous off-by-one fixes to remove confusing logic around subtracting an extra 1 here and there.

VladLazar reviewed (commented):
There's still some things I don't understand, so I'll need to come back to it.

graphcareful commented on src/v/raft/log_eviction_stm.h:
```diff
@@ -56,16 +56,18 @@ class log_eviction_stm {
     ss::gate _gate;
-    model::offset _previous_eviction_offset;
     model::offset _requested_eviction_offset;
+    std::deque<model::offset> _eviction_offsets;
```
Yup this was definitely a place I wanted feedback in. It doesn't hurt to do this, we will still process the events in the order they would have arrived in before these changes were implemented, but now there are other truncation events via delete-records requests as well.

IMO the strategy should be truncate at the max offset from all of the events that have been obtained. The smaller offsets can be thrown away.

graphcareful replied:
Yeah, I went with this method though because there would need to be no locking, and when an offset for an event that is lower then visible_offset is observed, there will be no work to be done.

graphcareful commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -16,23 +16,140 @@
+    /// before it should be truncated
+    auto offset = _raft->get_offset_translator_state()->to_log_offset(
+      kafka_offset);
+    const auto high_watermark = model::next_offset(_raft->last_visible_index());
```
Yeah there are a few checks, some already there in the log eviction stm today. Having another check here prevents replicating a special batch that when processed would be ignored anyway.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ebb4b29487e9c1de0dc3a5d5b811b34845800879..3a05354302677a3cd9797ee97f133e831f3cfb56)
- Fix for CI issues which were causing some retention based tests to fail. Issue was writing a snapshot with offset::min as the value of `log_start_delta`
- Fixed issue where truncation at bounds wasn't entirely respected, new ducktape test written to test for this case. (log of one record, truncate at 0 & 1)
- Expanded a ducktape test to truncates at the high watermark, only truncation at 1 before the high watermark was tested.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/3a05354302677a3cd9797ee97f133e831f3cfb56..2fb1f65d5732e2debda17219e292d99727b0b1d6)
- Reduce eviction housekeeping interval from 1s to 200ms. Fixed a failing raft unit test that had a 2s timeout on checking a segments eviction. 1s also seemed a bit too long, open to discussion on this.
- Removed a useless tracelog from `log_eviction_stm`

VladLazar commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -70,61 +219,131 @@ void log_eviction_stm::process_log_eviction() {
+            _eviction_offsets.pop_front();
+            _requested_eviction_offset = event.offset;
+            const auto log_start_offset = [this]() {
+                const auto ss_idx = _raft->last_snapshot_index();
```
Shouldn't this be model::next_offset(_raft->last_snapshot_index())? `_raft->last_snapshot_index()` is the truncation point. no?

graphcareful replied:
Yeah but we need to take into consideration the newly added delta. The delta represents how far into the last_snapshotted_(segment) the new start offset is. This is only non-zero in the case a user made a delete-records request between a segment boundary

VladLazar replied:
Yes, I guess i was asking why you are avoiding the `+1` since you want the first visible offset here.

NyaliaLui commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,238 @@
+        if (!cfg || !cfg->properties.cleanup_policy_bitflags) {
+            return false;
+        }
+        /// TODO: call is_collectable()? Maybe theres overrides
+        const auto flags = cfg->properties.cleanup_policy_bitflags;
+        return (*flags & model::cleanup_policy_bitflags::deletion)
+               == model::cleanup_policy_bitflags::deletion;
```
I agree that `storage::ntp_config::is_collectable()` seems to check the same thing; however, it looks like you'd have to create an `ntp_config` before you can call `is_collectable`. Perhaps, adding a `ntp_config` conversion within `topic_properties` or `topic_configuration` would make sense.

NyaliaLui commented on src/v/kafka/protocol/schemata/delete_records_request.json:
```diff
@@ -0,0 +1,42 @@
+    { "name": "TimeoutMs", "type": "int32", "versions": "0+",
+      "about": "How long to wait for the deletion to complete, in milliseconds." }
+  ]
+}
```
In the commit message, could you add the Kafka version that this is based on?
For example, see the commit message in https://github.com/redpanda-data/redpanda/pull/7685/commits/e7acf3c562000b4f710ba34a9810869220b07df1

graphcareful commented on src/v/raft/consensus.cc:
```diff
@@ -2232,20 +2248,31 @@ ss::future<> consensus::write_snapshot(write_snapshot_cfg cfg) {
-      _log);
+      last_start_delta() >= 0, "Log start offset delta cannot be less than 0");
+
+    // Support taking snapshots of one segment
```
Maybe the comment should be changed to refer to taking a snapshot when we are at the very first segment. It is a situation where `last_snapshot_index` would be == to `offset::min`. A snapshot would never be made unless there is another segment created, therefore the above call to get_term would only be called on a valid offset.

With delete-records we want to support moving the base ahead even 1 index at a time, even if there is only 1 segment present (the first) which means we will need to be able to write snapshots, even when `last_snapshot_index` is `offset::min`. This is needed to move the newly added log delta forward.

graphcareful commented on src/v/raft/log_eviction_stm.h:
```diff
@@ -56,18 +67,32 @@ class log_eviction_stm {
+    ss::future<model::offset> do_process_log_eviction(eviction_event);
+
+    ss::future<> apply(model::record_batch) override;
+    ss::future<> handle_eviction() override { return ss::now(); }
```
Good point. However is that necessary? This is the eviction STM, so it drives eviction events, could there ever be a case where `handle_eviction` is called?

VladLazar replied:
Yeah, imagine a case where an out of date node re-joins the cluster (perhaps after a long-ish connectivity issue). The STM was still spinning in that case. I think calling advancing `_next` here to the start offset is fine here, since the the special eviction batch must be replicated on all nodes. This means that we can't miss one such batch by advancing (if one node was isloated, replicating the batch would have failed).

I'm not sure about wheter we should advance to raft's start offset or raft's start offset plus the start offset delta added in this PR.

VladLazar replied:
`_next` can fall bellow the start offset if the node becomes isolated and re-joins later on. Local retention does not replicate a `prefix_truncate` batch, so I think that after a period of isolation where local retention kicked in, `state_machine::apply` will try to create a reader starting from offset `_next`. Since local retention moved the start offset forward and `_next` hasn't been updated as no batches were applied, `_next` will now be below the start offset.

Shouldn't be too hard to try out this scenario and see what happens.

graphcareful commented on src/v/raft/consensus.cc:
```diff
@@ -2232,20 +2248,31 @@ ss::future<> consensus::write_snapshot(write_snapshot_cfg cfg) {
+    // Support taking snapshots of one segment
+    auto last_included_term = std::optional<model::term_id>(model::term_id{});
+    if (last_included_index >= model::offset{0}) {
+        _log.get_term(last_included_index);
```
Leaving a note here, that changed fixed all remaining failures but broke the delete-records tests, so there is a bug in this logic here
```
ERROR 2023-04-20 16:19:10,230 [shard 1] assert - Assert failure: (/var/lib/buildkite-agent/builds/buildkite-amd64-builders-i-0cd3c19828bb13f0c-1/redpanda/redpanda/src/v/raft/consensus.cc:2275) 'last_included_term.has_value()' Unable to get term for snapshot last included offset: 10101, log: {offsets: {start_offset:10102, committed_offset:10243, committed_offset_term:1, dirty_offset:10243, dirty_offset_term:1, last_term_start_offset:10102}, is_closed: false, segments: [{size: 1, [{offset_tracker:{term:1, base_offset:10102, committed_offset:10243, dirty_offset:10243}, compacted_segment=0, finished_self_compaction=0, generation={12}, reader={/var/lib/redpanda/data/kafka/topic-gbablhblwr/0_18/10102-1-v1.log, (72840 bytes)}, writer={no_of_chunks:64, closed:0, fallocation_offset:1064960, committed_offset:72840, bytes_flush_pending:0}, cache={cache_size=6}, compaction_index:nullopt, closed=0, tombstone=0, index={file:/var/lib/redpanda/data/kafka/topic-gbablhblwr/0_18/10102-1-v1.base_index, offsets:{10102}, index:{header_bitflags:0, base_offset:{10102}, max_offset:{10243}, base_timestamp:{timestamp: 1682007549539}, max_timestamp:{timestamp: 1682007549540}, batch_timestamps_are_monotonic:1, index(2,2,2)}, step:32768, needs_persistence:1}}]}], config: {ntp: {kafka/topic-gbablhblwr/0}, base_dir: /var/lib/redpanda/data, overrides: {compaction_strategy: {nullopt}, cleanup_policy_bitflags: {delete}, segment_size: {nullopt}, retention_bytes: {}, retention_time_ms: {}, recovery_enabled: false, retention_local_target_bytes: {}, retention_local_target_ms: {}, remote_delete: {true}, segment_ms: {}}, revision: 18, initial_revision: 18}}
```

VladLazar commented:
/ci-repeat

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9dd1e604025c7197dc2952c6b04df29a8e0b87c8..8848941d30b96cca66074e5f37a38955c1aaa514)
- Fixed bug where last term was not properly initialized when writing snapshots that arose from delete-records events

graphcareful commented:
Change in [force-push](https://github.com/redpanda-data/redpanda/compare/15e97f6efb3994ed786224646907349f5cbadc37..187597f7ca3963a1003a4c70c36cfb942c5d535b)
- Fixed failure in `storage_resources_test.py` which correctly found a bug in the modifications made to `log_eviction_stm`. Bug was the `log_eviction_stm` reading from the start of the log and re-processing already committed offsets, fix is to set the offset at which the `log_eviction_stm` starts at to the `raft->committed_offset()`

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/187597f7ca3963a1003a4c70c36cfb942c5d535b..1189f7cc50cde38a366c9317a6b9ed3ea298c8d8)
- Removed `visible_start_offset` replaced with actually including new delta in the calculation for `start_offset`
- Implemented `handle_eviction` in `log_eviction_stm` to just move to first offset to process in this case.
- Implemented additional check to fail on prefix_truncation of cloud storage enabled topics at the cluster level

mmaslankaprv commented on src/v/raft/types.h:
```diff
@@ -511,18 +511,21 @@ struct transfer_leadership_options {
 };
 
 using offset_translator_delta = named_type<int64_t, struct ot_delta_tag>;
+using log_start_offset_delta = named_type<int64_t, struct lso_delta_tag>;
```
can this be `model::offset` ?

mmaslankaprv commented on src/v/raft/types.h:
```diff
@@ -511,18 +511,21 @@ struct transfer_leadership_options {
     group_configuration latest_configuration;
     ss::lowres_clock::time_point cluster_time;
     offset_translator_delta log_start_delta;
+    log_start_offset_delta last_start_delta;
```
Maybe we can simply store an absolute offset value and same it `effective_start_offset` ?

mmaslankaprv commented on src/v/raft/consensus.h:
```diff
@@ -331,8 +331,12 @@ class consensus {
     size_t received_snapshot_bytes() const { return _received_snapshot_bytes; }
     bool has_pending_flushes() const { return _has_pending_flushes; }
 
+    log_start_offset_delta last_snapshot_delta() const {
+        return _last_start_delta;
+    }
```
this name is confusing, i proposed storing an absolute value for an effective start offset, this way we would get rid of delta we can have then `log_start_offset` and `start_offset` or something similar ?

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -2185,20 +2195,26 @@ ss::future<> consensus::write_snapshot(write_snapshot_cfg cfg) {
                          // do nothing, we already have snapshot for this offset
                          // MUST be checked under the _op_lock
                          if (cfg.last_included_index <= _last_snapshot_index) {
-                             return ss::make_ready_future<bool>(false);
+                             if (cfg.last_start_delta <= _last_start_delta) {
```
`||` ?

mmaslankaprv commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -16,23 +16,161 @@
   ss::lw_shared_ptr<storage::stm_manager> stm_manager,
   ss::abort_source& as)
-  : _raft(raft)
+  : raft::state_machine(raft, logger, ss::default_priority_class())
```
do we want to have default priority class here ? Maybe it should depend from the partition type ?

mmaslankaprv commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -16,23 +16,161 @@
+static bool is_within_bounds(
+  ss::logger& logger,
+  ss::lw_shared_ptr<const storage::offset_translator_state> translator,
+  model::offset kafka_truncation_offset,
```
I think we should do the translation somewhere in the caller layer all raft operates on log offsets, this introduces a notion of something Kafka specific into that world. Maybe perform translation in `kafka::replicated_partition` ?

mmaslankaprv commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -16,23 +16,161 @@
+        /// officially been incremented at or above the requested truncated
+        /// offset on this node. No guarantees of data removal / availability
+        /// can be made at or after this point.
+        co_await _raft->min_visible_offset_monitor().wait(
+          as_log_offset, deadline, as);
```
it looks like we do not need the `min_visible_offset_monitor` the wait in `replicate_commands` should be enough.

graphcareful replied:
I did this because the `wait` in `replicate_commands` will return when the log end moves forward. This portion of code is waiting on the start offset to move forward. The start offset is already far behind the head so if I do this logic with `wait`, the command immediately returns and this will return to the client before the effect was actually performed.

mmaslankaprv commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -70,61 +210,135 @@ void log_eviction_stm::process_log_eviction() {
+                "segment_boundary_offset {}",
+                eviction_offset,
+                segment_boundary_offset);
+              return _raft
+                ->write_snapshot(write_snapshot_cfg(
+                  segment_boundary_offset,
+                  log_start_offset_delta(delta),
+                  iobuf()))
+                .then([eviction_offset] { return eviction_offset; });
```
What if we wouldn't store the `log_start_offset_delta` or `effective_start_offset` in `snapshot_metadata` but make it part of `iobuf()` this way we wouldn't have to change raft interface at all. the only part that would change is `cluster:partition` it would query the `_log_eviction_stm` for start offset i think that solution would be much cleaner as it wouldn't have to introduce delete offsets specific changes in general Raft algorithm implementation

mmaslankaprv commented:
I think the major question i have is: is it possible to encapsulate the delete offset logic in `eviction_stm` and not changing raft mechanisms ?

graphcareful commented:
> I think the major question i have is: is it possible to encapsulate the delete offset logic in `eviction_stm` and not changing raft mechanisms ?

Yes these latest changes have implemented the feedback suggested by Michal to do just this. I decided to go with the approach to remove the proposed `start_offset_delta` and just have the `consensus` class be able to write snapshots at any arbitrary index. This removes the necessity to modify any raft structs that were previously in this PR

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/e72b26d807266ba9a0eb2afd5d3a4dbb5fe7aef2..a93b0a3bfe253e33e58bcf0556f6d9242cb667c5)
- Removed start offset delta and any associated changes to raft types
- Simplified `log_eviction_stm` it simply calls `write_snapshot` with whatever offset it grabbed from an event
- Made `raft::consensus` able to write snapshots at any arbitrary index, by ensuring the offset translator does not truncate more state then is necessary when calling `offset_translator_state::prefix_truncate`
- Fixed a bug that was exists when crashing a node shortly after delete-records. Solution is to keep track of the last stm applied offset and only apply when write_snapshot is called
- Edited ducktape tests to introduce random crashes.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2417983f3a3761931906358ba01ba606c8988af4..19d415f0081a0c412b657736a72394496074c47a)
- Introduced a bug in previous changes that invoked `state_machine::write_last_applied` the intention was to store a snapshot but it was incorrectly implemented.
- Modified implementation to have `log_eviction_stm` inherit from `persistent_stm` which gives an stm the power to store arbitrary state in snapshots.
- This necessitated the move of `persistent_stm` from the `cluster` layer to the `raft` layer, which was simple due to the fact that it did not depend on anything in the `cluster` layer anyway.

piyushredpanda commented:
/ci-repeat 1

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f76535b809db425b2207c8ad6bde49743a5791d4..e813c9d13cfc6e8261663e7c21f1884fbd27fb84)
- Rebase dev

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/e813c9d13cfc6e8261663e7c21f1884fbd27fb84..56094b058750385c5eeb9f737d5899e73e44a388)
- Ensure new log_eviction_stm snapshots are cleaned up when partition is deleted. Was causing some CI tests to fail.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/95d7db2772ef9c3f9fd666f658552b495bab167c..dc114f145951a0dee9e1ccbdd59058337917854e)
- Fixed a bug where eviction stm didn't exit early on writing stale snapshots

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/235b7b0b8a23743eb408905c6f293ef207b9b6bf..cff1f819d95c4af08c76bbd75d3ae15848cc7051)
- Rebased against dev

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/3c52d0ff4dc161dc4d0c56ae4d656b0480d90a4e..0fc3255e1f283471db8c1108436b29084d480680)
- Fixed a bug where start offset could move forward without writing snapshot.

graphcareful commented:
/ci-repeat 4 skip-unit release dt-repeat=32

mmaslankaprv commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -60,53 +61,70 @@ void log_eviction_stm::monitor_log_eviction() {
+    } catch (const ss::sleep_aborted&) {
+        vlog(_logger.trace, "sleep aborted");
+    } catch (const std::exception& ex) {
+        vlog(_logger.warn, "Exception encountered: {}", ex);
```
can we add some context to those messages ?

mmaslankaprv commented on src/v/raft/log_eviction_stm.cc:
```diff
@@ -60,53 +61,70 @@ void log_eviction_stm::monitor_log_eviction() {
+    try {
+        if (_eviction_offsets.empty()) {
+            co_await ss::sleep_abortable(200ms, _as);
+        } else {
```
can we use a condition variable here instead of sleep ?

mmaslankaprv commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1308,6 +1308,22 @@ disk_log_impl::get_term_last_offset(model::term_id term) const {
+        return std::nullopt;
+    }
+    auto& idx = (*it)->index();
+    if (auto entry = idx.find_nearest(o)) {
```
i think this will not work as expected. We use sparse index i.e. the nearest batch may not be the one closest to the offset but the one closest to the offset and indexed.

graphcareful replied:
If there is no way around this the expected behavior is that the actual data removed on disk will be slightly greater then if we did know the nearest segment boundary to the requested deletion offset, which is expected because a user could request to delete at any arbitrary index.

Opinions?

dotnwat replied:
>actual data removed on disk will be slightly greater

Wouldn't we want to avoid removing any data at offsets higher than requested, and if the offset were within a batch or in the middle of a segment then we'd do a logical truncation? What does kafka compatibility say about it?

mmaslankaprv replied:
if we will stick to the current semantics of this method we must at least rename it to reflect the behavior

mmaslankaprv replied:
@graphcareful i think this comment is still not addressed

graphcareful replied:
Thanks for reminding me, i think we agreed on a rename, would `nearest_known_batch_boundary_offset` make sense?

graphcareful replied:
Renamed to `index_lower_bound`

graphcareful commented:
/ci-repeat 4 skip-unit release dt-repeat=4

graphcareful commented:
Only CI failure is known: https://github.com/redpanda-data/redpanda/issues/10868

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/0fc3255e1f283471db8c1108436b29084d480680..6bd3819645f2e68aa37a32c4e07f504bb2d36814)
- Fixed a bug where eviction_stm was not properly respecting the max collectible offset threshold
- Added mechanism to prevent infinite looping in the eviction_stm when reprocessing events occurs
- Added new test that invokes DeleteRecords in the middle of an open transaction

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/6bd3819645f2e68aa37a32c4e07f504bb2d36814..c6deba4cfe04894a9a63f54c4e87d9813af18f33)
- Fixed missing bits in consensus layer to properly read stored raft snapshots
- Storing effective start offset in the raft snapshot

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c6deba4cfe04894a9a63f54c4e87d9813af18f33..f68fa291161151b30ad2ca365a9daeaf0bd2c83b)
- Rebased dev to resolve conflict with `feature_table.h`

dotnwat commented on src/v/raft/errc.h:
```diff
@@ -39,6 +39,7 @@ enum class errc : int16_t {
     replicate_batcher_cache_error,
     group_not_exists,
     replicate_first_stage_exception,
+    invalid_truncation_offset
```
what does it mean for an offset to be invalid?

graphcareful replied:
out of range

dotnwat reviewed (commented):
do the prefix truncate batches need to be preserved after a compaction event?

graphcareful commented:
> do the prefix truncate batches need to be preserved after a compaction event?

Yeah , nice point, i think this exposes a bug, that the special batches have similar keys (iobuf()) and the data is stored in the value portion of the record. Swapping these would solve the issue.

graphcareful commented:
> Wouldn't we want to avoid removing any data at offsets higher than requested, and if the offset were within a batch or in the middle of a segment then we'd do a logical truncation? What does kafka compatibility say about it?

Did you see that from a comment? It must be incorrect then, your assumption is right we definitely don't want to remove data after the requested offset. I probably meant to write "actual data removed on disk will be slightly less"

dotnwat commented:
> > do the prefix truncate batches need to be preserved after a compaction event?
> 
> Yeah , nice point, i think this exposes a bug, that the special batches have similar keys (iobuf()) and the data is stored in the value portion of the record. Swapping these would solve the issue.

I think that you should exempt the batch type from compaction (safer, more explicit semantics), rather than trying to trick the compaction algorithm.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b7a988f57e58dd89490c09f14ad701d9e4a801a8..03584ecc49d699f267c7c1e8f0c2493157531011)
- Fix bug that raised assertion in `effective_start_offset` , now set new start offset before writing snapshot so it can never be the case that `effective_start` < `raft->last_snapshot`
- Adressed Noah's comment about compaction compacting away prefix batches. Modified the special batch type data to go into the key field of a record instead of the value field. This way only identical requests to prefix truncate will be compacted away

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/03584ecc49d699f267c7c1e8f0c2493157531011..8b5bf74c92f5cae9eb5dae5d991bd0668654d31c)
- Rebased dev

mmaslankaprv commented on src/v/cluster/log_eviction_stm.h:
```diff
@@ -32,18 +35,31 @@ class consensus;
  * adjust that offset with _stm_manager->max_collectible_offset(), write the
  * raft snapshot and notify the storage layer that log eviction can safely
  * proceed up to the adjusted offset.
+ *
```
This commit would benefit from more detailed commit message, it changes multiple interfaces and abstractions would be great to know what is the reasoning

mmaslankaprv commented on src/v/cluster/partition.cc:
```diff
@@ -170,6 +176,32 @@ partition::partition(
+ss::future<std::error_code> partition::prefix_truncate(
+  model::offset truncation_offset, ss::lowres_clock::time_point deadline) {
+    if (!_log_eviction_stm) {
+        throw std::runtime_error(
+          "Partition not prefix-truncatable, retention settings not applied");
```
maybe return an error code instead of throwing ?

mmaslankaprv commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -46,6 +46,15 @@ ss::future<> log_eviction_stm::start() {
+    /// recent known version of the special batch. This is particularly useful
+    /// to know if the start offset is up to date in the case leadership has
+    /// recently changed for example.
+    static const auto sync_effective_start_timeout = 5s;
```
maybe it would be better to pass in the timeout, this way we may change it depending on the situation

mmaslankaprv commented on src/v/cluster/log_eviction_stm.h:
```diff
@@ -79,6 +79,10 @@ class log_eviction_stm final
 
+    /// Ensure followers have processed up until the most recent known version
+    /// of the batch representing the start offset
+    ss::future<bool> sync_effective_start();
```
nit: can this return `result<model::offset>` ?

mmaslankaprv commented on src/v/kafka/server/handlers/list_offsets.cc:
```diff
@@ -106,6 +106,15 @@ static ss::future<list_offset_partition_response> list_offsets_partition(
+        // the leader it will not return a stale result for start_offset
+        auto synced = co_await kafka_partition->sync_start_offset();
+        if (!synced) {
+            co_return list_offsets_response::make_partition(
+              ktp.get_partition(), error_code::unknown_server_error);
```
Same here with an error

mmaslankaprv commented on src/v/kafka/server/handlers/list_offsets.cc:
```diff
@@ -106,6 +106,15 @@ static ss::future<list_offset_partition_response> list_offsets_partition(
      * that the actual timestamp be returned. only the offset is required.
      */
     if (timestamp == list_offsets_request::earliest_timestamp) {
+        // using sync_start checks that all peers have processed the most known
```
actually the sync verifies if the leader processed all updates

mmaslankaprv commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,345 @@
+}
+
+ss::future<std::error_code> log_eviction_stm::truncate(
+  model::offset kafka_truncate_offset,
```
would it be possible for us to perform the translation in `kafka` layer i.e. replicated_partition, we generally stick to do that and only exceptionally do it in other places. In this case it seems that we can use translated offset here

graphcareful replied:
This is sometime during the handling of the request, so maybe it would belong in `kafka/handler/delete-records.cc` or `cluster/partition.cc`?

andrwng replied:
I suspect we will run into trouble doing it in the Kafka layer, given this stm needs to know about the kafka offset to pass it to the archival layer, which isn't able to do offset translation. I.e., this call needs to work with an offset that doesn't exist in local data.

mmaslankaprv commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,345 @@
+      kafka_truncate_offset,
+      _raft->last_snapshot_index(),
+      _raft->last_visible_index());
+    auto result = co_await replicate_commands(std::move(batch), deadline, as);
+    if (result) {
+        co_return result;
+    }
```
This logic is pretty tricky, we ignore an error and move to waiting, maybe it would be good to place a comment here and log the result with debug log ?

mmaslankaprv commented on src/v/cluster/log_eviction_stm.h:
```diff
@@ -32,40 +35,84 @@ class consensus;
+    /// Signaled when a snapshot should be taken, and data deleted
+    ss::condition_variable _reap_condition;
+    /// Signaled when the start offset moves forward
+    raft::offset_monitor _min_consumable_offset_monitor;
```
is this start offset monitor ?

mmaslankaprv commented on src/v/cluster/partition.cc:
```diff
@@ -170,6 +176,40 @@ partition::partition(
+    }
+    const auto within_bounds = [this](model::offset kafka_truncation_offset) {
+        const auto& translator = _raft->get_offset_translator_state();
+        const auto kafka_start_offset = translator->from_log_offset(
```
i would move this validation to `kafka::replicated_partition`

mmaslankaprv commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+    const auto is_deletable = [&ctx](const delete_records_topic& t) {
+        const auto cfg = ctx.metadata_cache().get_topic_cfg(
+          model::topic_namespace_view(model::kafka_namespace, t.name));
+        if (!cfg || !cfg->properties.cleanup_policy_bitflags) {
```
i think in case there is no `cfg` wee should return `invalid_topic_exception`

mmaslankaprv commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+
+    if (!is_authorized(t)) {
+        return make_partition_errors(
+          t, error_code::cluster_authorization_failed);
```
should that be `topic_authorization_failed` ?

mmaslankaprv commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+        const auto si_flags = cfg->properties.shadow_indexing;
+        return si_flags && *si_flags != model::shadow_indexing_mode::disabled;
+    };
+    const auto is_internal_topic = [](const delete_records_topic& t) {
+        const auto& internal_topics
+          = config::shard_local_cfg().kafka_nodelete_topics();
+        return std::find_if(
+                 internal_topics.begin(),
+                 internal_topics.end(),
+                 [t](const ss::sstring& name) { return name == t.name; })
+               != internal_topics.end();
+    };
```
Here naming is a bit inconsistent, i.e. method is named `is_internal_topic` but the check validates if a topic is on non delete list

mmaslankaprv commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+              = ctx.partition_manager()
+                  .invoke_on(
+                    *shard,
+                    [ntp,
```
move ?

mmaslankaprv commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+    auto errc = co_await partition->prefix_truncate(
+      truncation_offset, ss::lowres_clock::now() + timeout_ms);
+    if (errc) {
+        vassert(
+          errc.category() == cluster::error_category(),
+          "Unknown error category observed");
```
Do we really want to crash in this case ? Can we see Raft/RPC errors ?

mmaslankaprv commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+          kerr);
+        co_return make_partition_error(ntp, kerr);
+    }
+    const auto kafka_start_offset
+      = partition->get_offset_translator_state()->from_log_offset(
+        partition->start_offset());
```
Using replicated partition would provide consistent interface it would be enough to call `start_offset` method.

mmaslankaprv commented on src/v/cluster/log_eviction_stm.h:
```diff
@@ -83,6 +84,11 @@ class log_eviction_stm final
 
+    /// Ensure followers have processed up until the most recent known version
+    /// of the batch representing the start offset
+    ss::future<result<model::offset, cluster::errc>>
```
i think we shouldn't limit ourselves to `cluster::errc` simple `result<model::offset>` will handle errors coming from all the layers then we can differentiate them using error category.

mmaslankaprv commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -77,8 +78,12 @@ static ss::future<read_result> read_from_partition(
     if (unlikely(!lso)) {
         co_return read_result(lso.error());
     }
+    auto maybe_start_ofs = co_await part.sync_effective_start();
+    if (!maybe_start_ofs) {
+        co_return read_result(map_topic_error_code(maybe_start_ofs.error()));
+    }
```
I think this should be part of `replicated_partition::validate_fetch_offset`, this way we would validate the start offset with updated value, here it is only returned like this to the client

andrwng commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,345 @@
+        /// officially been incremented at or above the requested truncated
+        /// offset on this node. No guarantees of data removal / availability
+        /// can be made at or after this point.
+        co_await _min_consumable_offset_monitor.wait(
+          as_log_offset, deadline, as);
```
Do we need this `wait()` anymore, given the only `notify()` is in the `apply()` call, which IIUC should be waited for anyway in `replicate_commands()` at L206?

Once the `apply()` is complete, further fetches will fail since we take the requested offset into account when looking at `start_offset()` in `replicated_partition::validate_fetch_offset()`

graphcareful replied:
Resolving, removed the `start_offset_monitor` and replaced with `wait_no_throw`

andrwng commented on src/v/kafka/server/handlers/delete_records.cc:
```diff
@@ -0,0 +1,248 @@
+    const auto is_deletable = [&ctx](const delete_records_topic& t) {
+        const auto cfg = ctx.metadata_cache().get_topic_cfg(
+          model::topic_namespace_view(model::kafka_namespace, t.name));
+        if (!cfg || !cfg->properties.cleanup_policy_bitflags) {
```
[is_collectable](https://github.com/redpanda-data/redpanda/blob/dev/src/v/storage/ntp_config.h#L134) also checks for the cleanup_policy_bitflags, but handles things differently (returns that it is deletable if there is none set). I'm not sure, but do we need to copy that logic? I see this causing trouble in fixture tests, but without digging further I'm not sure it's a problem outside of tests

graphcareful replied:
Yeah i would saw we should operate the same way other consumers of that API would react

graphcareful commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,345 @@
+    /// snapshots as close to effective_start_offset as possible.
+    auto gh = _gate.hold();
+    while (!_as.abort_requested()) {
+        try {
+            /// Five seconds is the most the condition variable would sleep for,
+            /// in the common path condvar.signal() will wake up the condition
+            /// variable and data will be reaped as soon as the incoming event
+            /// increments the start offset.
+            static const auto reap_timeout_sec = std::chrono::seconds(5);
+            co_await _reap_condition.wait(reap_timeout_sec);
```
I agree, I don't think we even need the while, the next event to wakeup the fiber will attempt to reap everything up until the new start.

mmaslankaprv commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,347 @@
+              "Condition variable broken, shutting down background data "
+              "reaping fiber");
+            break;
+        } catch (const ss::condition_variable_timed_out&) {
+            ;
+        }
```
nit: this is no longer required

mmaslankaprv commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,347 @@
+    if (batch.header().type == model::record_batch_type::prefix_truncate) {
+        /// record_batches of type ::prefix_truncate are always of size 1
+        const auto truncate_offset = serde::from_iobuf<model::offset>(
+          batch.copy_records().begin()->key().copy());
```
we can skip last copy here and release the key

mmaslankaprv commented on src/v/cluster/log_eviction_stm.cc:
```diff
@@ -9,104 +9,347 @@
+    ///
+    /// Here all apply() does is move forward the in memory start offset, a
+    /// background fiber is responsible for evicting as much as possible
+    if (batch.header().type == model::record_batch_type::prefix_truncate) {
```
nit: unlikely

graphcareful commented:
CI failures
- https://github.com/redpanda-data/redpanda/issues/11593
- https://github.com/redpanda-data/redpanda/issues/11455
- https://github.com/redpanda-data/redpanda/issues/11480

Investigating a leak seen in `RackAwarePlacementTest.test_replica_placement`

graphcareful commented:
/ci-repeat 1
debug
skip-units
dt-repeat=20
tests/rptest/tests/rack_aware_replica_placement_test.py::RackAwarePlacementTest.test_replica_placement

graphcareful commented:
/ci-repeat 1
debug
skip-units
dt-repeat=200
tests/rptest/tests/memory_stress_test.py::MemoryStressTest.test_fetch_with_many_partitions
